ported (#6730)
GoogleChrome/web.dev: src/site/content/en/blog/building-a-3d-game-menu-component/index.md

---
layout: post
title: Building a 3D game menu component
subhead: A foundational overview of how to build a responsive, adaptive, and accessible 3D game menu.
authors:
  - adamargyle
description: A foundational overview of how to build a responsive, adaptive, and accessible 3D game menu.
date: 2021-11-10
hero: image/vS06HQ1YTsbMKSFTIPl2iogUQP73/IjOV7KRfEzEj7I1ZVJA5.png
thumbnail: image/vS06HQ1YTsbMKSFTIPl2iogUQP73/1yLFCkKYvDtrJSepc7A0.png
tags:
  - blog
  - css
  - dom
  - javascript
---

In this post I want to share thinking on a way to build a 3D game menu component. Try the
[demo](https://gui-challenges.web.app/game-menu/dist/).

<figure class="w-figure w-figure--fullbleed">
  {% Video
    src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/q9gIXMmP2CcfutwyaK5F.mp4",
    class="w-screenshot",
    autoplay="true",
    loop="true",
    muted="true"
  %}
  <figcaption class="w-figure">
    <a href="https://gui-challenges.web.app/game-menu/dist/">Demo</a>
  </figcaption>
</figure>

If you prefer video, here's a YouTube version of this post:

{% YouTube 'x0aCdLDZ8BM' %}

## Overview

> Video games often present users with a menu that is super different, animated

Reviewer: ```suggestion
Video games often present users with a creative and unusual menu, animated
```

that demonstrates the 3D space the card is within, even if a mouse can't or
isn't interacting with the component.

```css
@media (--motionOK) {
  .threeD-button-set {
    /* browser hint so it can be prepared and optimized */
    will-change: transform;

    /* transition transform style changes and run an infinite animation */
    transition: transform .1s ease;
    animation: rotate-y 5s ease-in-out infinite;
  }
}
```

The `rotate-y` animation only sets the middle keyframe at `50%` since the
browser will default `0%` and `100%` to the default style of the element. This
is shorthand for animations that alternate, needing to begin and end at the same
position. It's a great way to articulate infinite alternating animations.

```css
@keyframes rotate-y {
  50% {
    transform: rotateY(15deg) rotateX(-6deg);
  }
}
```

{% Video
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/EYqJVUyybb3VV4AX9dNE.mp4",
  class="w-screenshot",
  autoplay="true",
  loop="true",
  muted="true"
%}

### Styling the `<li>` elements

Each list item (`<li>`) contains the button and its border elements. The
`display` style is changed so the item doesn't show a

> [`::marker`](https://web.dev/css-marker-pseudo-element/). The `position` style

Reviewer: ```suggestion
[`::marker`](/css-marker-pseudo-element/). The `position` style
```

pad or joysticks to move focus on a gamepad. The
[roving-ux](https://github.com/argyleink/roving-ux) library often used for GUI
Challenge interfaces will handle arrow keys for us. The below code tells the
library to trap focus within `.threeD-button-set` and forward the focus to the
button children.

```js
import {rovingIndex} from 'roving-ux'

rovingIndex({
  element: document.querySelector('.threeD-button-set'),
  target: 'button',
})
```

{% Video
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/PSF8lpd2tcmj35WIQlKT.mp4",
  class="w-screenshot",
  autoplay="true",
  loop="true",
  muted="true"
%}

### Mouse parallax interaction

Tracking the mouse and having it tilt the menu is intended to mimic AR and VR
video game interfaces, where instead of a mouse you may have a virtual pointer.
It can be fun when elements are hyper aware of the pointer.

{% Video
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/mNtHI14zZMoX250VIIZF.mp4",
  class="w-screenshot",
  autoplay="true",
  loop="true",
  muted="true"
%}

Since this is a small extra feature, we'll put the interaction behind a query of
the user's motion preference. Also, as part of setup, store the button list

> component into memory with `querySelector` and cache the element’s bounds into

Reviewer: ```suggestion
component into memory with `querySelector` and cache the element's bounds into
```
